### DISCUSSION TITLE
Transakcja

### DISCUSSION THREAD

**Discussion by M-Szymczyk** [M-Szymczyk/ProjektPortfel#19] [General]
Klasa AmountOfMoney implementuje interfejs AmountOfMoneyInterface. w tym interefesie zadeklarowałeś metode trensaction() do operowania na transakcjach. W tej metodzie pytałeś czy jest wpłata czy wypłata gotówki. Jako, że usuwam wszystkie zapytania z pakietu data to można przenieś to zapytanie do aplikacji i przesłać do tej metody ten wybór jako parametr, albo jako enum albo można też uznać, że ujemny wpływ gotówki to wypłata.

**Comment by wolski0420** [comment]
Podawanie w argumencie czy to wypłata czy wpłata to raczej herezja, to nie C żeby flagi przekazywać xDD Proponuję dać po prostu metody withdraw() i deposit()w AmountOfMoney, a wykrywanie transakcji zrobić na wyższej warstwie tj. Window App, Console App itd. Czyli z poziomu użytkownika jesteśmy najpierw w klasie np. MoneyChangerConsoleApp, tam dokonuje się decyzja czy wypłacamy czy wpłacamy i wtedy w zależności od wyboru wywoływana jest jedna z tych dwóch metod na niższej warstwie. Co Ty na to?

**Reply by M-Szymczyk** [reply]
Zrobię tak jak proponujesz

**Reply by M-Szymczyk** [reply]
Nie przesyłanie jak to nazwałeś flagi skutkuje sporym powtarzaniem kodu, więc wolałem to czasem przesłać niż przeklejać kod
Chociaż teraz też kod jest powtarzany w niektórych miejscach

**Reply by wolski0420** [reply]
W sumie to nie bardzo widzę te powtórzenia :P coś może by się znalazło ale nie ma tragedii, w pakiecie Data raczej jest całkiem ok

**Reply by wolski0420** [reply]
Można w sumie przenieść te funkcje na obiekty. Jak? Przydatny może być Command Pattern lub coś zbliżonego. Interfejs Command posiadałby metodę pod np. nazwą "compute" która przyjmowała by np. dwa argumenty, jeden to nasz budżet, drugi to kwota która ma być odjęta/dodana do naszego budzetu (bardzo bardzo ułatwione, dzięki BigDecimalom a nie intom, wystarczy referencja w parametrze, nie trzeba nic zwracać). Następnie dwie klasy implementujące ten interfejs odpowiednio: Withdraw do odejmowania, Deposit do dodawania. W jednym compute odejmuje argumenty, w drugim dodaje i wszystko bryka. CommandExecutorem nie wiem co by jeszcze było, czy oddzielna klasa czy gdzieś w istniejącej jakiejś ale myślę że nie byłby to duży problem. Jakie z tego korzyści? Więcej obiektowości, mniej funkcyjności przede wszystkim, możliwe też że nie będzie jakichś powtórzeń bo ten pattern potrafi dość uporządkować projekty, po za tym może ułatwić w przyszłości wykonywanie stałych przelewów które będzie trzeba realizować za pewne przez Timery, a te działają dość fajnie z Command Patternem ;)

**Reply by wolski0420** [reply]
Gdyby to miało nastąpić to mogę się tym zająć, czekam jednak najpierw na finalnego pull requesta z Twojej strony
